### ISSUE TITLE
High CPU by FreeSwitch

### ISSUE THREAD

**Issue by amguirado73** [bigbluebutton/bigbluebutton#9472]
Hello,

We have stressed a Bigbluebutton server through 100  "listen-only" users generated by a python script that uses Selenium. We have verified that these users generate channels against Freeswitch and this process is consuming a lot of CPU to establish the conference between them. According to the documentation, Listen-only users should receive the audio via kurento (streaming) and release Freeswitch to do this job. Could you confirm this?. Is there something in the configuration to change?

Thanks.

**Comment by MrCapitalist** [comment]
Hi
I had high CPU usage problem too. With 60 users my server was freezed and load average was increased to 38!! (with 12 core CPU)
After changing RAM of server, it retured to normal and now load average is 4 with 60 users

**Comment by ffdixon** [comment]
We've been tracking a related issue in FreeSWITCH where it will occasionally stop accepting a websocket request, which can cause a 1002 error for the user.  A user can re-attempt to connect and it usually works.

However, the bbb-webrtc-sfu also needs to connect to FreeSWITCH to establish the Listen Only connection.  If that connections fails due to FreeSWITCH not accepting the request at that moment, then all subsequent request for Listen Only will transfer from Kurento to FreeSWITCH, and the user will connect to FreeSWITCH with a full two-way channel that is muted (so to the user it looks like Listen Only).

We've been working on fixing this underlying websocket issue in FreeSWITCH with the core FreeSWITCH team and have been making some progress.  We'll update this ticket as we make more progress.

**Comment by ffdixon** [comment]
@amguirado73 Could you do the test again with 100 Listen Only users, but this time disable the jitterbuffer in `/opt/freeswitch/conf/dialplan/default/bbb_conference.xml` and `bbb_echo_to_conference.xml` and let us know if you get a similar CPU spike.

**Comment by amguirado73** [comment]
Hello,

we are going to the test today. We have simply deleted the jitterbuffer 
line in both xml files. Is it ok?.

Thanks

**Comment by amguirado73** [comment]
Hi,

we have done the following changes in both xml configuration files (comment out the directives): 



Now, we can have 200 users with a 60% used CPU (we send the graph). Our server is a VM with 
8 CORES (Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz) and 16 GB RAM.

We have simulated a 197 only-listens users (2 new users each 60 second) and 3 moderators users without video. We can speak fluently with no delays.

We can't understand why a Listen-Only user must create a channel in the freeswitch conference instead of using Kurento, because is very expensive in CPU, but now we can achieve 200 users with 60% CPU.

Thanks you very much for your support.

[test.pdf](https://github.com/bigbluebutton/bigbluebutton/files/4609470/test.pdf)

**Comment by prlanzarin** [comment]
@amguirado73 There should be one listen only channel per meeting (and a **few** ones that usually come from users falling back to FS listen only if the GLOBAL_AUDIO one failed).
If you're having that many channels for a single meeting, something is wrong with your configuration (most probably).

Can you please send the output of `sudo netstat -putan | grep -i 5066` and the content of
`/usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml` `freeswitch.sip_ip` field?
Please change the IP addresses for something else if you consider it sensitive info. But state if they're local of public IPs.

**Comment by amguirado73** [comment]
Hello,

here you are:

cat /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml | grep sip
  sip_ip: 192.168.200.245


netstat -putan | grep -i 5066
tcp        0      0 80.73.144.135:5066      0.0.0.0:*               LISTEN      16753/freeswitch
tcp        0      1 80.73.144.135:41192     192.168.200.245:5066    SYN_SENT    17394/node
tcp6       0      0 ::1:5066                :::*                    LISTEN      16753/freeswitch

Thank you very much for your support.
Regards

**Comment by prlanzarin** [comment]
Hmm. It's wrong.
Change `/usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml` -> `freeswitch.sip_ip` to `80.73.144.135`, then `sudo systemctl restart bbb-webrtc-sfu` and try it again.

**Comment by amguirado73** [comment]
Thank you. We have changed it and launched new tests.

**Comment by amguirado73** [comment]
Hello,

we have done new test and now we can have 200 users with 20% CPU. Very good performance. Now we have simulated a meeting with 4 VM clients with python and selenium generated request. but we have reduced the time between request from 60 seconds to 30 seconds (simulating that user connect quickly).  Now I think we have the problem that @ffdixon wrote. After 150-200 request, the users start to be identified as LISTEN-ONLY-username and channels in FREESWITCH grow:

uuid,direction,created,created_epoch,name,state,cid_name,cid_num,ip_addr,dest,application,application_data,dialplan,context,read_codec,read_rate,read_bit_rate,write_codec,write_rate,write_bit_rate,secure,hostname,presence_id,presence_data,accountcode,callstate,callee_name,callee_num,callee_direction,call_uuid,sent_callee_name,sent_callee_num,initial_cid_name,initial_cid_num,initial_ip_addr,initial_dest,initial_dialplan,initial_context
73fe31b7-d250-49bb-879b-87d675aae308,inbound,2020-05-11 22:23:29,1589228609,sofia/external/w_0paiylyuizfj_2-bbbID-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_0paiylyuizfj_2-bbbID-Antonio,w_0paiylyuizfj_2-bbbID-Antonio,80.73.144.135,919672013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_0paiylyuizfj_2-bbbID-Antonio,w_0paiylyuizfj_2-bbbID-Antonio,80.73.144.135,919672013,XML,public
d97d8f5e-a2d4-4570-9ccc-4035486aa9a9,inbound,2020-05-11 22:33:17,1589229197,sofia/external/GLOBAL_AUDIO_72013@80.73.144.135,CS_EXECUTE,GLOBAL_AUDIO_72013,GLOBAL_AUDIO_72013,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,,BigBlueButton,,,,ACTIVE,,,,,,,GLOBAL_AUDIO_72013,GLOBAL_AUDIO_72013,80.73.144.135,72013,XML,public
3d5237e3-47af-4047-bb4a-2a7aa2151dea,inbound,2020-05-11 23:16:59,1589231819,sofia/external/w_z13soaaroxk4_1-bbbID-LISTENONLY-EspiaB41@bigbluebutton.f-integra.org,CS_EXECUTE,w_z13soaaroxk4_1-bbbID-LISTENONLY-EspiaB41,w_z13soaaroxk4_1-bbbID-LISTENONLY-EspiaB41,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_z13soaaroxk4_1-bbbID-LISTENONLY-EspiaB41,w_z13soaaroxk4_1-bbbID-LISTENONLY-EspiaB41,80.73.144.135,72013,XML,public
a9af0a87-fdf6-409e-9401-5497dae2c70d,inbound,2020-05-11 23:18:53,1589231933,sofia/external/w_lermcwlrfrya_1-bbbID-LISTENONLY-EspiaB42@bigbluebutton.f-integra.org,CS_EXECUTE,w_lermcwlrfrya_1-bbbID-LISTENONLY-EspiaB42,w_lermcwlrfrya_1-bbbID-LISTENONLY-EspiaB42,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_lermcwlrfrya_1-bbbID-LISTENONLY-EspiaB42,w_lermcwlrfrya_1-bbbID-LISTENONLY-EspiaB42,80.73.144.135,72013,XML,public
ffd86783-23cd-4332-8987-ef6f0d599144,inbound,2020-05-11 23:21:32,1589232092,sofia/external/w_tovsuw5muuzr_1-bbbID-LISTENONLY-EspiaB44@bigbluebutton.f-integra.org,CS_EXECUTE,w_tovsuw5muuzr_1-bbbID-LISTENONLY-EspiaB44,w_tovsuw5muuzr_1-bbbID-LISTENONLY-EspiaB44,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_tovsuw5muuzr_1-bbbID-LISTENONLY-EspiaB44,w_tovsuw5muuzr_1-bbbID-LISTENONLY-EspiaB44,80.73.144.135,72013,XML,public
906bf50b-1cc6-406c-8d10-187f27d753b5,inbound,2020-05-11 23:23:08,1589232188,sofia/external/w_ix6xwzlzaply_1-bbbID-LISTENONLY-EspiaA33@bigbluebutton.f-integra.org,CS_EXECUTE,w_ix6xwzlzaply_1-bbbID-LISTENONLY-EspiaA33,w_ix6xwzlzaply_1-bbbID-LISTENONLY-EspiaA33,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_ix6xwzlzaply_1-bbbID-LISTENONLY-EspiaA33,w_ix6xwzlzaply_1-bbbID-LISTENONLY-EspiaA33,80.73.144.135,72013,XML,public

We think that at the beginning of the meeting, the number of requests will be very high, and if the problem is reproduced, the number of channels on the freeswitch will grow, and therefore the CPU consumption.

We are considering using an intermediate page with a random timeout (up to 5 minutes) to distribute to users. However, we are concerned about this incidence. 

Thanks

**Comment by noonscoomo** [comment]
amguirado73, if you provide the python script we would join the testing. We see similar issues in our school-installations. It seems that randomly some people can't get in with micro and we were currently thinking about a reproducible load test.

**Comment by ffdixon** [comment]
Update `bbb-install.sh` to properly set `freeswitch.sip_ip` in `default.yml`, see: https://github.com/bigbluebutton/bbb-install/commit/61fab4905509d281ab332b1197659ebfff285f12#diff-632a52fe9625496439841de370d44a8cR921

**Comment by levigo-systems** [comment]
@ffdixon 

> Could you do the test again with 100 Listen Only users, but this time disable the jitterbuffer in /opt/freeswitch/conf/dialplan/default/bbb_conference.xml and bbb_echo_to_conference.xml and let us know if you get a similar CPU spike.

is this recommended on a productive environment or should this not be implemented?

**Comment by foobarable** [comment]
@ffdixon Could you elaborate why setting the internal IP will fix this issue?

**Comment by amguirado73** [comment]
Hi, 

we have checked that python and selenium are very heavy to do a 300-400 users test. Now we have used a simple script .bat in windows:

for /L %%b in (1,1,50) do (
ECHO %i
firefox -new-tab "https://bigbluebutton.yourdoamin/bigbluebutton/api/join?meetingID=XXXX&fullName=User&password=XXXX&guest=false&checksum=5dafa4a7dd4c9e7c2e5a1af66e9961616657b39853ffbb7ae0a67da935519c95"
timeout 5
)

This URL can be obtained  using developing tools in chrome:

https://developers.google.com/web/tools/chrome-devtools/open

When you join to your meeting, you can see the URL to the API join action . You can copy the URL and use in any browser. We have created a meeting with a lock to create only-listener user (lockSettingsDisableMic = "true")

In a windows VM, you create a .bat in the same folder where firefox is, you can generate a lot of requests.

Now we are generating 50 users to reproduce the freeswitch issue. For example in the last run we have the following 16 channels in freeswitch:

freeswitch@BigBlueButton> show channels
uuid,direction,created,created_epoch,name,state,cid_name,cid_num,ip_addr,dest,application,application_data,dialplan,context,read_codec,read_rate,read_bit_rate,write_codec,write_rate,write_bit_rate,secure,hostname,presence_id,presence_data,accountcode,callstate,callee_name,callee_num,callee_direction,call_uuid,sent_callee_name,sent_callee_num,initial_cid_name,initial_cid_num,initial_ip_addr,initial_dest,initial_dialplan,initial_context
78ea35f3-4f04-4dc2-a82c-f9e278b02fad,inbound,2020-05-13 08:34:39,1589351679,sofia/external/w_oeiprrcvczaw_1-bbbID-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_oeiprrcvczaw_1-bbbID-Antonio,w_oeiprrcvczaw_1-bbbID-Antonio,80.73.144.135,919672013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_oeiprrcvczaw_1-bbbID-Antonio,w_oeiprrcvczaw_1-bbbID-Antonio,80.73.144.135,919672013,XML,public
bc7a8d65-d7ad-46bb-b7fa-bc1dd989141b,inbound,2020-05-13 10:25:41,1589358341,sofia/external/GLOBAL_AUDIO_72013@80.73.144.135,CS_EXECUTE,GLOBAL_AUDIO_72013,GLOBAL_AUDIO_72013,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,768000,opus,48000,0,,BigBlueButton,,,,ACTIVE,,,,,,,GLOBAL_AUDIO_72013,GLOBAL_AUDIO_72013,80.73.144.135,72013,XML,public
9a7e3950-7880-45c8-b1e1-9557fd1a419f,inbound,2020-05-13 10:28:32,1589358512,sofia/external/w_x5dy3q3es2ul_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_x5dy3q3es2ul_1-bbbID-LISTENONLY-Antonio,w_x5dy3q3es2ul_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_x5dy3q3es2ul_1-bbbID-LISTENONLY-Antonio,w_x5dy3q3es2ul_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
91a710f7-60be-4700-bae8-86eb96e8c66c,inbound,2020-05-13 10:28:35,1589358515,sofia/external/w_hjq076mxjauq_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_hjq076mxjauq_1-bbbID-LISTENONLY-Antonio,w_hjq076mxjauq_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_hjq076mxjauq_1-bbbID-LISTENONLY-Antonio,w_hjq076mxjauq_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
8ca390d8-0600-495e-98d0-6696e64f4bd3,inbound,2020-05-13 10:29:17,1589358557,sofia/external/w_dlq0y2pssotg_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_dlq0y2pssotg_1-bbbID-LISTENONLY-Antonio,w_dlq0y2pssotg_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_dlq0y2pssotg_1-bbbID-LISTENONLY-Antonio,w_dlq0y2pssotg_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
b61b5780-6500-4a62-969f-da27ffcb8351,inbound,2020-05-13 10:30:04,1589358604,sofia/external/w_goq7tqrd7qrh_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_goq7tqrd7qrh_1-bbbID-LISTENONLY-Antonio,w_goq7tqrd7qrh_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_goq7tqrd7qrh_1-bbbID-LISTENONLY-Antonio,w_goq7tqrd7qrh_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
7797c3fe-372a-458a-970b-897cdc72e87b,inbound,2020-05-13 10:30:09,1589358609,sofia/external/w_1gluczqlmo5u_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_1gluczqlmo5u_1-bbbID-LISTENONLY-Antonio,w_1gluczqlmo5u_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_1gluczqlmo5u_1-bbbID-LISTENONLY-Antonio,w_1gluczqlmo5u_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
fb484d94-e0c0-42c8-b8eb-36b37a9b6e66,inbound,2020-05-13 10:30:33,1589358633,sofia/external/w_c2ufzvreuzen_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_c2ufzvreuzen_1-bbbID-LISTENONLY-Antonio,w_c2ufzvreuzen_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_c2ufzvreuzen_1-bbbID-LISTENONLY-Antonio,w_c2ufzvreuzen_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
c8f74bde-ae3c-4bba-b4c7-a6997edbc0d1,inbound,2020-05-13 10:30:40,1589358640,sofia/external/w_iz4vkjnbrvok_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_iz4vkjnbrvok_1-bbbID-LISTENONLY-Antonio,w_iz4vkjnbrvok_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_iz4vkjnbrvok_1-bbbID-LISTENONLY-Antonio,w_iz4vkjnbrvok_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
d261cafd-a23a-4f6a-8b8e-3af0327251e3,inbound,2020-05-13 10:31:21,1589358681,sofia/external/w_d5xrp2lvisx5_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_d5xrp2lvisx5_1-bbbID-LISTENONLY-Antonio,w_d5xrp2lvisx5_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_d5xrp2lvisx5_1-bbbID-LISTENONLY-Antonio,w_d5xrp2lvisx5_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
03be2da1-9fbe-46c7-81f1-1857a1e4d175,inbound,2020-05-13 10:32:10,1589358730,sofia/external/w_a4vwrtz8xl3h_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_a4vwrtz8xl3h_1-bbbID-LISTENONLY-Antonio,w_a4vwrtz8xl3h_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_a4vwrtz8xl3h_1-bbbID-LISTENONLY-Antonio,w_a4vwrtz8xl3h_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
173674e6-92d9-4f9f-b28c-a1b84f51cf8d,inbound,2020-05-13 10:33:50,1589358830,sofia/external/w_oiti7rg76wei_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_oiti7rg76wei_1-bbbID-LISTENONLY-Antonio,w_oiti7rg76wei_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_oiti7rg76wei_1-bbbID-LISTENONLY-Antonio,w_oiti7rg76wei_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
52278a54-fad9-48cd-b0c0-094a4b2c500f,inbound,2020-05-13 10:34:34,1589358874,sofia/external/w_cwi1u6bhc1s5_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_cwi1u6bhc1s5_1-bbbID-LISTENONLY-Antonio,w_cwi1u6bhc1s5_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_cwi1u6bhc1s5_1-bbbID-LISTENONLY-Antonio,w_cwi1u6bhc1s5_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
a5052450-92aa-40e5-bf4c-b016070fd6b1,inbound,2020-05-13 10:34:39,1589358879,sofia/external/w_jhpmlnw4m5kd_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_jhpmlnw4m5kd_1-bbbID-LISTENONLY-Antonio,w_jhpmlnw4m5kd_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_jhpmlnw4m5kd_1-bbbID-LISTENONLY-Antonio,w_jhpmlnw4m5kd_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
032748a8-fbd8-4d9f-887c-34dba55e9fbf,inbound,2020-05-13 10:34:46,1589358886,sofia/external/w_shwmi3cmutfk_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_shwmi3cmutfk_1-bbbID-LISTENONLY-Antonio,w_shwmi3cmutfk_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_shwmi3cmutfk_1-bbbID-LISTENONLY-Antonio,w_shwmi3cmutfk_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public
c93f6cd3-1b0a-4c09-b3d3-c53790980210,inbound,2020-05-13 10:38:08,1589359088,sofia/external/w_sc3gajduilgk_1-bbbID-LISTENONLY-Antonio@bigbluebutton.f-integra.org,CS_EXECUTE,w_sc3gajduilgk_1-bbbID-LISTENONLY-Antonio,w_sc3gajduilgk_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_sc3gajduilgk_1-bbbID-LISTENONLY-Antonio,w_sc3gajduilgk_1-bbbID-LISTENONLY-Antonio,80.73.144.135,72013,XML,public

16 total.

**Comment by amguirado73** [comment]
Hi again, 

we have done 60 users performances test wit the .bat script using a windows VM. I have noticed that after 53-54 users, all new users use FREESWITCH and they aren't subcribed to KURENTO global stream. I send you log information:

- A user that use FREESWITCH has id: w_piadostn8r0b

show channels
82698402-c473-4234-9c38-ee321906e2f8,inbound,2020-05-13 13:28:13,1589369293,sofia/external/w_piadostn8r0b_1-bbbID-LISTENONLY-User@bigbluebutton.f-integra.org,CS_EXECUTE,w_piadostn8r0b_1-bbbID-LISTENONLY-User,w_piadostn8r0b_1-bbbID-LISTENONLY-User,80.73.144.135,72013,conference,72013@cdquality,XML,default,L16,48000,1536000,opus,48000,0,srtp:dtls:AES_CM_128_HMAC_SHA1_80,BigBlueButton,,,,ACTIVE,,,,,,,w_piadostn8r0b_1-bbbID-LISTENONLY-User,w_piadostn8r0b_1-bbbID-LISTENONLY-User,80.73.144.135,72013,XML,public


- Here are the logs from bbb-webrtc-sfu.log:

{"roomId":"72013","userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] New user created at room 72013 with id cool-mountain-8Dt1Ioq","timestamp":"2020-05-13T11:20:03.545Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","userId":"w_ozxtvmszore2","userName":"User","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Starting new listen only session","timestamp":"2020-05-13T11:20:03.546Z"}
{"level":"info","message":"[mcs-sdp-session] New session created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"mediaSession\",\"mediaSessionId\":\"throbbing-snowflake-LBKjKoS\",\"mediaId\":\"throbbing-snowflake-LBKjKoS\",\"medias\":[],\"roomId\":\"72013\",\"userId\":\"cool-mountain-8Dt1Ioq\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"mediaTypes\":{\"video\":false,\"audio\":false,\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"isMuted\":false,\"volume\":50,\"strategy\":\"freewill\"}","timestamp":"2020-05-13T11:20:03.549Z"}
{"level":"info","message":"[mcs-sdp-media] New media created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"media\",\"mediaSessionId\":\"throbbing-snowflake-LBKjKoS\",\"mediaId\":\"hidden-field-AXedA7i\",\"roomId\":\"72013\",\"userId\":\"cool-mountain-8Dt1Ioq\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"muted\":false,\"volume\":50,\"talking\":false,\"mediaTypes\":{\"video\":false,\"audio\":\"recvonly\",\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"subscribedTo\":\"\",\"strategy\":\"freewill\",\"adapterElementId\":null}","timestamp":"2020-05-13T11:20:03.551Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Listen only session successfully subscribed to global audio with MCS user cool-mountain-8Dt1Ioq","timestamp":"2020-05-13T11:20:03.612Z"}
{"level":"info","message":"[AudioManager] Started listen only session for user w_ozxtvmszore2 at 72013 with connectionId 109","timestamp":"2020-05-13T11:20:03.612Z"}
{"level":"info","message":"[AudioManager] Sending startResponse to user w_ozxtvmszore2 at 72013 with connectionId 109","timestamp":"2020-05-13T11:20:03.613Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"elementId":"9c793b11-5653-4ab8-a9ca-fa5d44e3302d_kurento.MediaPipeline/8707543b-2a77-4f1a-b58c-7a4c2bdf5513_kurento.WebRtcEndpoint","level":"info","message":"[audio] Audio ICE gathering done","timestamp":"2020-05-13T11:20:03.665Z"}

{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"elementId":"9c793b11-5653-4ab8-a9ca-fa5d44e3302d_kurento.MediaPipeline/8707543b-2a77-4f1a-b58c-7a4c2bdf5513_kurento.WebRtcEndpoint","iceComponentId":1,"localCandidate":"candidate:16 1 UDP 2013266426 80.73.144.135 27441 typ host","remoteCandidate":"candidate:remote1 1 UDP 1853817087 80.73.144.141 59782 typ prflx","level":"info","message":"[audio] Audio new candidate pair selected","timestamp":"2020-05-13T11:20:03.680Z"}

- From bbb-webrtc-sfu.log (cat bbb-webrtc-sfu.log | grep ozxtvmszore2). You can see that there is a Audio new candidate pair selected message and the user is changed from kurento to freeswitch:

{"roomId":"72013","userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] New user created at room 72013 with id cool-mountain-8Dt1Ioq","timestamp":"2020-05-13T11:20:03.545Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","userId":"w_ozxtvmszore2","userName":"User","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Starting new listen only session","timestamp":"2020-05-13T11:20:03.546Z"}
{"level":"info","message":"[mcs-sdp-session] New session created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"mediaSession\",\"mediaSessionId\":\"throbbing-snowflake-LBKjKoS\",\"mediaId\":\"throbbing-snowflake-LBKjKoS\",\"medias\":[],\"roomId\":\"72013\",\"userId\":\"cool-mountain-8Dt1Ioq\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"mediaTypes\":{\"video\":false,\"audio\":false,\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"isMuted\":false,\"volume\":50,\"strategy\":\"freewill\"}","timestamp":"2020-05-13T11:20:03.549Z"}
{"level":"info","message":"[mcs-sdp-media] New media created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"media\",\"mediaSessionId\":\"throbbing-snowflake-LBKjKoS\",\"mediaId\":\"hidden-field-AXedA7i\",\"roomId\":\"72013\",\"userId\":\"cool-mountain-8Dt1Ioq\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"muted\":false,\"volume\":50,\"talking\":false,\"mediaTypes\":{\"video\":false,\"audio\":\"recvonly\",\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"subscribedTo\":\"\",\"strategy\":\"freewill\",\"adapterElementId\":null}","timestamp":"2020-05-13T11:20:03.551Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Listen only session successfully subscribed to global audio with MCS user cool-mountain-8Dt1Ioq","timestamp":"2020-05-13T11:20:03.612Z"}
{"level":"info","message":"[AudioManager] Started listen only session for user w_ozxtvmszore2 at 72013 with connectionId 109","timestamp":"2020-05-13T11:20:03.612Z"}
{"level":"info","message":"[AudioManager] Sending startResponse to user w_ozxtvmszore2 at 72013 with connectionId 109","timestamp":"2020-05-13T11:20:03.613Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"elementId":"9c793b11-5653-4ab8-a9ca-fa5d44e3302d_kurento.MediaPipeline/8707543b-2a77-4f1a-b58c-7a4c2bdf5513_kurento.WebRtcEndpoint","level":"info","message":"[audio] Audio ICE gathering done","timestamp":"2020-05-13T11:20:03.665Z"}

{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"elementId":"9c793b11-5653-4ab8-a9ca-fa5d44e3302d_kurento.MediaPipeline/8707543b-2a77-4f1a-b58c-7a4c2bdf5513_kurento.WebRtcEndpoint","iceComponentId":1,"localCandidate":"candidate:16 1 UDP 2013266426 80.73.144.135 27441 typ host","remoteCandidate":"candidate:remote1 1 UDP 1853817087 80.73.144.141 59782 typ prflx","level":"info","message":"[audio] Audio new candidate pair selected","timestamp":"2020-05-13T11:20:03.680Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Listen only WebRTC media is FLOWING","timestamp":"2020-05-13T11:20:04.598Z"}

{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Sending global audio connection for user w_ozxtvmszore2","timestamp":"2020-05-13T11:20:04.599Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"elementId":"9c793b11-5653-4ab8-a9ca-fa5d44e3302d_kurento.MediaPipeline/8707543b-2a77-4f1a-b58c-7a4c2bdf5513_kurento.WebRtcEndpoint","mediaState":"CONNECTED","level":"info","message":"[audio] Audio media state changed","timestamp":"2020-05-13T11:20:09.988Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":109,"level":"info","message":"[audio] Listen only WebRTC media is CONNECTED","timestamp":"2020-05-13T11:20:09.989Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":"109","level":"info","message":"[audio] Sending global audio disconnection for user w_ozxtvmszore2","timestamp":"2020-05-13T11:29:48.126Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"cool-mountain-8Dt1Ioq","userId":"w_ozxtvmszore2","userName":"User","mediaId":"throbbing-snowflake-LBKjKoS","status":"MEDIA_STARTED","connectionId":"109","level":"info","message":"[audio] Removing user with connectionId 109","timestamp":"2020-05-13T11:29:48.126Z"}
{"userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","roomId":"72013","level":"info","message":"[mcs-user] User cool-mountain-8Dt1Ioq has autoLeave on and no medias, will be ejected","timestamp":"2020-05-13T11:29:48.131Z"}
{"userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] User cool-mountain-8Dt1Ioq  wants to leave","timestamp":"2020-05-13T11:29:48.131Z"}
{"userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] User cool-mountain-8Dt1Ioq  actually leaving","timestamp":"2020-05-13T11:29:48.131Z"}
{"userId":"cool-mountain-8Dt1Ioq","externalUserId":"w_ozxtvmszore2","roomId":"72013","level":"info","message":"[mcs-user] User cool-mountain-8Dt1Ioq has autoLeave on and no medias, will be ejected","timestamp":"2020-05-13T11:29:48.131Z"}
{"roomId":"72013","userId":"nameless-sun-r0fifIX","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] New user created at room 72013 with id nameless-sun-r0fifIX","timestamp":"2020-05-13T11:30:17.924Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","userId":"w_ozxtvmszore2","userName":"User","status":"MEDIA_STARTED","connectionId":157,"level":"info","message":"[audio] Starting new listen only session","timestamp":"2020-05-13T11:30:17.924Z"}
{"level":"info","message":"[mcs-sdp-session] New session created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"mediaSession\",\"mediaSessionId\":\"green-surf-Xi_mm~_\",\"mediaId\":\"green-surf-Xi_mm~_\",\"medias\":[],\"roomId\":\"72013\",\"userId\":\"nameless-sun-r0fifIX\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"mediaTypes\":{\"video\":false,\"audio\":false,\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"isMuted\":false,\"volume\":50,\"strategy\":\"freewill\"}","timestamp":"2020-05-13T11:30:17.927Z"}
{"level":"info","message":"[mcs-sdp-media] New media created {\"type\":\"WebRtcEndpoint\",\"memberType\":\"media\",\"mediaSessionId\":\"green-surf-Xi_mm~_\",\"mediaId\":\"bitter-bird-OK95lmp\",\"roomId\":\"72013\",\"userId\":\"nameless-sun-r0fifIX\",\"name\":\"bigbluebutton|subscribe|audio|w_ozxtvmszore2|8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260\",\"muted\":false,\"volume\":50,\"talking\":false,\"mediaTypes\":{\"video\":false,\"audio\":\"recvonly\",\"text\":false,\"content\":false,\"application\":false,\"message\":false},\"subscribedTo\":\"\",\"strategy\":\"freewill\",\"adapterElementId\":null}","timestamp":"2020-05-13T11:30:17.929Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"nameless-sun-r0fifIX","userId":"w_ozxtvmszore2","userName":"User","mediaId":"green-surf-Xi_mm~_","status":"MEDIA_STARTED","connectionId":157,"level":"info","message":"[audio] Listen only session successfully subscribed to global audio with MCS user nameless-sun-r0fifIX","timestamp":"2020-05-13T11:30:17.955Z"}
{"level":"info","message":"[AudioManager] Started listen only session for user w_ozxtvmszore2 at 72013 with connectionId 157","timestamp":"2020-05-13T11:30:17.955Z"}
{"level":"info","message":"[AudioManager] Sending startResponse to user w_ozxtvmszore2 at 72013 with connectionId 157","timestamp":"2020-05-13T11:30:17.956Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"nameless-sun-r0fifIX","userId":"w_ozxtvmszore2","userName":"User","mediaId":"green-surf-Xi_mm~_","status":"MEDIA_STARTED","connectionId":157,"elementId":"773fea30-0d81-44e2-9cb9-df0741651e38_kurento.MediaPipeline/c244692d-5fbf-47f5-9824-451baac68b42_kurento.WebRtcEndpoint","level":"info","message":"[audio] Audio ICE gathering done","timestamp":"2020-05-13T11:30:17.996Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"nameless-sun-r0fifIX","userId":"w_ozxtvmszore2","userName":"User","mediaId":"green-surf-Xi_mm~_","status":"MEDIA_STARTED","connectionId":157,"elementId":"773fea30-0d81-44e2-9cb9-df0741651e38_kurento.MediaPipeline/c244692d-5fbf-47f5-9824-451baac68b42_kurento.WebRtcEndpoint","iceComponentId":1,"localCandidate":"candidate:16 1 UDP 2013266426 80.73.144.135 28141 typ host","remoteCandidate":"candidate:remote1 1 UDP 1853817087 80.73.144.141 50775 typ prflx","level":"info","message":"[audio] Audio new candidate pair selected","timestamp":"2020-05-13T11:30:27.505Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"nameless-sun-r0fifIX","userId":"w_ozxtvmszore2","userName":"User","mediaId":"green-surf-Xi_mm~_","status":"MEDIA_STARTED","connectionId":157,"level":"info","message":"[audio] Sending global audio disconnection for user w_ozxtvmszore2","timestamp":"2020-05-13T11:30:29.966Z"}
{"roomId":"72013","internalMeetingId":"8d6cbd049c22493b1c4aa836d5bb7c45c3cd3044-1589366319260","mcsUserId":"nameless-sun-r0fifIX","userId":"w_ozxtvmszore2","userName":"User","mediaId":"green-surf-Xi_mm~_","status":"MEDIA_STARTED","connectionId":157,"level":"info","message":"[audio] Removing user with connectionId 157","timestamp":"2020-05-13T11:30:29.966Z"}
{"userId":"nameless-sun-r0fifIX","externalUserId":"w_ozxtvmszore2","roomId":"72013","level":"info","message":"[mcs-user] User nameless-sun-r0fifIX has autoLeave on and no medias, will be ejected","timestamp":"2020-05-13T11:30:29.967Z"}
{"userId":"nameless-sun-r0fifIX","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] User nameless-sun-r0fifIX  wants to leave","timestamp":"2020-05-13T11:30:29.968Z"}
{"userId":"nameless-sun-r0fifIX","externalUserId":"w_ozxtvmszore2","level":"info","message":"[mcs-controller] User nameless-sun-r0fifIX  actually leaving","timestamp":"2020-05-13T11:30:29.968Z"}
{"userId":"nameless-sun-r0fifIX","externalUserId":"w_ozxtvmszore2","roomId":"72013","level":"info","message":"[mcs-user] User nameless-sun-r0fifIX has autoLeave on and no medias, will be ejected","timestamp":"2020-05-13T11:30:29.969Z"}

**Comment by amguirado73** [comment]
I am very sorry if I am very repetitive. We are seriously considering adopting BBB as a solution for webimars and therefore we are doing the performance tests. In the previous posts we have commented that from a number of users the audio stopped being served by KURENTO and started to be served by FREESWICH (with active mute). After many tests, I'm not sure if the problem is with the server or the client that generated the tests. Is it possible that if a client was very slow to respond, bbb-webrtc-sfu decided to switch the audio to FREESWITCH?
If this were so, there really would be no problem and I would simply have to add more VMs in parallel to generate the tests.

Thank you very much.
Regards.

**Comment by prlanzarin** [comment]
Did you use a single VM for the 60 users test?

**Comment by amguirado73** [comment]
Hello,

yes, I have used only one VM for 60 users. 

Regards

**Comment by prlanzarin** [comment]
I don't know how big is your VM, but 60 is too much. I'd say a safe limit is about 20 if they're only listening/viewing.

**Comment by amguirado73** [comment]
Hello,

Tomorrow I will try to use more VMs to do the tests.

Thank you.
Regards

**Comment by gufmar** [comment]
We see similar things here
There are two main processes for Kurento and Freeswitch eating up a lot of CPU resources. 
And then quite some freeswitch processes at 1-2% of CPU usage. 
This is a VM with 8 vCores on a Dual Xeon Gold 2.2 GHz Server.

![bbb_htop_CPU-usage](https://user-images.githubusercontent.com/13198220/81915987-92196f80-95d3-11ea-9d7c-a015d224ee96.png)

As you can see in this resource monitoring over time CPU usage goes more ore less 1:1 with participants count 

![2020-05-14_11h15_10](https://user-images.githubusercontent.com/13198220/81916530-38fe0b80-95d4-11ea-87b6-76edef5da84d.png)

**Comment by gufmar** [comment]
> We've been working on fixing this underlying websocket issue in FreeSWITCH with the core FreeSWITCH team and have been making some progress. We'll update this ticket as we make more progress.

Does this mean, this underlying websocket issue also causes the high CPU usage. or is freeswitch generally generating a high CPU load for these bidirectional voice streams?

We tried to remove the jitterbuffer and restarted the server, but couldn't see big differences.

**Comment by amguirado73** [comment]
Hi,

after some time we can have done new tests. We have used 2 ESX (24 vCPU each one) with 6 Windos10 VM (4 vCPU and 8 GB RAM). Each VM could generate 20 clients, than consumed a lot of CPU. So we have generated 120 users Listen-Only without problems. No calls in freeswitch (only GLOBAL audio), no delay and CPU in BBB host was 10-15% CPU. So, BBB is fine to use in an production enviroremnt. The CPU in our ESX's were 100%, so we can not generate more users.

So, we very glad to use BBB.

Thank you very much.

**Comment by prlanzarin** [comment]
@gufmar Ignore the WebSocket issue and jitterbuffer for now. Follow what I suggested @amguirado73 to do in https://github.com/bigbluebutton/bigbluebutton/issues/9472#issuecomment-626755027.

- `sudo netstat -putan | grep -i 5066 | grep -i LISTEN`
- Get the IP FreeSWITCH is listening to 5066 on
- Set it in /usr/local/bigbluebutton/bbb-webrtc-sfu/config/default.yml freeswitch.sip_ip
- Restart bbb-webrtc-sfu.

See if the load goes down.

**Comment by gufmar** [comment]
@prlanzarin thank you
we already checked this. our servers public IP was already set in the default.yml
We use SSL (wss) over nginx port 7433. Not sure if this causes these CPU loads.

We tried a bit by playing around with sip/freeswtch/ws but so far no luck

![image](https://user-images.githubusercontent.com/13198220/82366364-6272c900-9a12-11ea-9a6f-d1abc5554064.png)

The question is also if our expectations for concurrent users are simply too high?  
8 vCores on a  Xeon Gold 2.2GHz Host seems to deal with up to 90 Users. 
Or is what we see here only affecting single Installations (for whatever reason) ?

**Comment by prlanzarin** [comment]
@gufmar Could you clarify some stuff on that scenario? Video means video publishers (cameras being shared)? And voice participants: are your users joining with mic or listen only in their majority?

**Comment by amguirado73** [comment]
Hello,

**@gufmar** our performance tests were with **"Listen-Only users" (moderator locked audio)**, no video and 1-2 moderators. So, CPU consumption was due to KURENTO. We also SSL (wss) over nginx port 7433. Our BBB version is 2.2.5 on Ubuntu 16.04.6 LTS. You can connect to Freeswitch with /opt/freeswitch/bin/fs_cli and check your audio channels (show channels). List-Only user must use streaming produced by KURENTO. So you should have a GLOBAL audio channel in Freeswitch (beside moderators channels).

Regards

**Comment by gufmar** [comment]
@prlanzarin this is a recording of metrics collected from the BBB API (https://docs.bigbluebutton.org/dev/api.html#getmeetinginfo)

repost it here with 1:1 using the BBB API key names:
![image](https://user-images.githubusercontent.com/13198220/82536625-078fbd80-9b49-11ea-845d-e20875badf83.png)
In the graph we sum up all current rooms counters and overlay it to the server CPU load.

The curves in this graph are not created by test scripts generating certain predefined loads. Instead, we have real users (students) in virtual classrooms, acting individually by enabling or switching their audio/video channels. Usually, most participants select bidirectional but muted audio. So they can quick answer questions from teachers on demand. 

What makes me wonder is, one such bidirectional channel seems to require ~1% from an 8core Xeon CPU (or alternatively counted as 8% of one Xeon core)

Would it make sense to set up a physical server and see if CPU usage is better then?

**Comment by gufmar** [comment]
but might I'm on the wrong track to look at audio channels.
We now have the resources captured in 10sec steps.

![image](https://user-images.githubusercontent.com/13198220/82544400-5ba09f00-9b55-11ea-9602-9511ea07e91e.png)

The most aligned lines at this 10sec resolution seem to be videoCount<=>CPU

Note: we configured 30kb Video as default, with an option to select 50kb

**Comment by beatwiz** [comment]
We run 10 dedicated 16c Xeon Servers and we observe that the CPU goes up when the number of Video Channels increases. Quality doesnt seem to have any performance impact (we have it fixed at 50kbps). 

We've had 10 meetings on a single server with 30 people each, and the server load was below 50%. 

There seems to be a direct relation between the number of rooms that have bi-directional video and server load. And when there are slow computers on the equation, that cannot handle more than, lets say, 15 webcams. They not only start to drag on the client side, but at the same time, they seem to drag the server aswell. As usually theres a spike of load on the server until the user timouts and disconnects. FWIW I have never actually measured this except with my eyes...

**Comment by basisbit** [comment]
@beatwiz your problem will be that the mostly single-threaded NodeJS/Meteor process for event handling, whiteboard, chat und participant list will hit 100% of one CPU core, and then the event processing plus the sessions on the client side just fail (slow response of the HTML5 frontend UI even in small sessions is quite a good indication for that). There are a bunch of improvements planned for (probably) bbb 2.3 that will improve this.
Also, please bear in mind that depending on the CPU architecture and linux kernel, freeswitch, NodeJS and Kurento might completely avoid virtual HyperThreading cores because using these would not help for dealing with more workload, thus causing your server to maybe already have hit the load limit at 50 to 60% of "total" CPU usage. Looking at the Linux Load Averages is usually the best indication. If it is higher than number of physical CPU cores, the your system hit the load limit.

**Comment by einhirn** [comment]
I know that there were some new releases since this issue was opened, but we're running 2.2.14 on our BBB Cluster and encountered a similar issue on 2 (of 12) separate Servers apart from each other. FreeSWITCH terminated some random (?) audio connections and didn't accept **any** new ones. Nginx error log shows "timeout connecting to upstream", netstat shows "syn_sent".

The remaining sessions still seemed to work ok, this only impacted new sessions.
Only thing that helped, eventually, was a restart of BBB. 

We're thankful for Scalelite because we were able to just "disable" the broken node and wait for the still working sessions to end...

**Comment by matiasilva** [comment]
Hm, this might the FreeSWITCH error that @basisbit has stumbled upon in the past.

**Comment by basisbit** [comment]
@einhirn your problem is actually this one: #9667 Also, existing sessions die shortly after the freeswitch websocket is gone because of missing heartbeat responses on the websocket port.

Anyways, let's close this issue here because the "high CPU usage by FreeSWITCH" was either just normal cpu load, missing firewall rules -> brute force attacks or ddos, or kurento being stalled and then FreeSWITCH handling all connections (tracked in other issues).

**Comment by matiasilva** [comment]
@basisbit There was an issue mentioned here: https://github.com/bigbluebutton/bigbluebutton/issues/9472#issuecomment-626755027 that I'm not sure if has been resolved

**Comment by prlanzarin** [comment]
It was.

**Comment by matiasilva** [comment]
Oh I linked the wrong comment. I meant the one mentioned by Fred a few comments above that one.

**Comment by prlanzarin** [comment]
If you're talking about the 1002 error, it hasn't, but it has its own issue (which basisbit linked). And it doesn't relate to high CPU usage in relation to its consequences.
The jitterbuffer thing Fred mentioned also has nothing to do with either CPU usage or websocket stalls.

**Comment by hmahdi-s** [comment]
@MrCapitalist 
could you please explain how to test hardware and fix?

**Comment by MrCapitalist** [comment]
I have high CPU load problem by FreeSwitch and mpeg process too. My server has 4*2.6GHz CPU cores + 8 GB DDR4 RAM + SSD storage. But when the number of users reaches 60, the CPU load increases and the speed decreases. So the new user can hardly enter the room >_<

**Comment by hmahdi-s** [comment]
@MrCapitalist 
I think what document says is not practical information.
It says with minimum server you can support 150 users but nobody confirm it in a practical use.
Maybe need some optimization...

**Comment by basisbit** [comment]
@MrCapitalist please do not hijack such old and already answered + closed issues in the issue tracker for support requests. 
Instead, please use the [forum / mailing list](https://groups.google.com/forum/#!forum/bigbluebutton-setup). This issue tracker here is only for BBB development related issues. Sorry.

**Comment by shabsshab** [comment]
I am getting very high CPU utilization(338.2% of CPU ) , how i can reduce this 

Users  around : 350
Server : Dedicated Server Quad core+HT ( Intel i7 , 8 Core Processor ) , 64GB DDR4 RAM, NVMe SSD.

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
20496 freeswi+  -2  19 4602200 1.153g  15784 S 338.2  1.8 594:27.21 freeswitch
20594 kurento   20   0 14.590g 1.432g  23808 S  33.9  2.3 256:41.79 kurento-media-s


root@BBBServer ~ # netstat -putan | grep -i 5066
tcp        0      0 138.201.x.x:5066     0.0.0.0:*               LISTEN      20496/freeswitch
tcp        0      0 138.201.x.x:55088    138.201.x.x:5066     ESTABLISHED 21192/node
tcp        0      0 138.201.x.x:5066     138.201.x.x:55088    ESTABLISHED 20496/freeswitch
tcp6       0      0 2a01:4f8:172:999:::5066 :::*                    LISTEN      20496/freeswitch
root@BBBServer ~ #

**Comment by basisbit** [comment]
@shabsshab please do not comment on old and already closed issues. Also, that CPU usage is not very high, it is to be expected / normal.
